mixer/pkg/protobuf/yaml/dynamic/builder.go

```diff
@@ -0,0 +1,280 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dynamic
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/gogo/protobuf/protoc-gen-gogo/descriptor"
+
+	"istio.io/api/policy/v1beta1"
+	"istio.io/istio/mixer/pkg/lang/compiled"
+	"istio.io/istio/mixer/pkg/protobuf/yaml"
+)
+
+type (
+
+	// Compiler compiles expression and returns a value type.
+	Compiler interface {
+		// Compile type Compiler interface {compiles expression and returns a value type.
+		Compile(text string) (compiled.Expression, v1beta1.ValueType, error)
+	}
+
+	// Builder builds encoder based on data
+	Builder struct {
+		resolver    yaml.Resolver
+		compiler    Compiler
+		skipUnknown bool
+	}
+)
+
+// NewEncoderBuilder creates an EncoderBuilder.
+func NewEncoderBuilder(resolver yaml.Resolver, compiler Compiler, skipUnknown bool) *Builder {
+	return &Builder{
+		resolver:    resolver,
+		compiler:    compiler,
+		skipUnknown: skipUnknown}
+}
+
+// Build builds a DynamicEncoder
+func (c Builder) Build(msgName string, data map[interface{}]interface{}) (Encoder, error) {
+	m := c.resolver.ResolveMessage(msgName)
+	if m == nil {
+		return nil, fmt.Errorf("cannot resolve message '%s'", msgName)
+	}
+
+	return c.buildMessage(m, data, true)
+}
+
+func (c Builder) buildMessage(md *descriptor.DescriptorProto, data map[interface{}]interface{}, skipEncodeLength bool) (Encoder, error) {
+	var err error
+	var ok bool
+
+	me := messageEncoder{
+		skipEncodeLength: skipEncodeLength,
+	}
+	isMap := md.GetOptions().GetMapEntry()
+
+	keys := make([]string, 0, len(data))
+	for kk := range data {
+		var k string
+		if k, ok = kk.(string); !ok {
+			return nil, fmt.Errorf("error processing message '%s':%v got %T want string", md.GetName(), kk, kk)
+		}
+		keys = append(keys, k)
+	}
+	// Ensure consistent log output.
+	sort.Strings(keys)
+
+	for _, k := range keys {
+		v := data[k]
+		fd := yaml.FindFieldByName(md, k)
+		if fd == nil {
+			if c.skipUnknown {
+				continue
+			}
+			return nil, fmt.Errorf("field '%s' not found in message '%s'", k, md.GetName())
+		}
+
+		switch fd.GetType() {
+		case descriptor.FieldDescriptorProto_TYPE_STRING:
+			var ma []interface{}
+			if fd.IsRepeated() {
+				if ma, ok = v.([]interface{}); !ok {
+					return nil, fmt.Errorf("unable to process: %v, got %T, want: []interface{}", fd, v)
+				}
+			} else {
+				ma = []interface{}{v}
+			}
+
+			for _, vv := range ma {
+				var fld *field
+				if fld, err = c.buildPrimitiveField(vv, fd, isMap && k == "key"); err != nil {
```

Comment:
should we follow the same pattern as messageField and do `me.fields = append(me.fields, fld)` inside it?